### PR TITLE
Revamp server-led mechanism

### PR DESCRIPTION
**Pull request by jefferai**
This replaces the previous mechanism of generating the full credential set and returning it with creating and storing an activation token that can be passed in as the nonce in a fetch request.

This allows the mechanism on the node side to be essentially the same, including all of the private keys being generated node-side. Instead of generating and signing a nonce on its own, it signs the activation token as the nonce.

### CODE REVIEW

File: registration/register_node_led.go [hunk 1]
```diff
@@ -230,12 +311,29 @@ func AuthorizeNode(
...
+// Note: this is called via paths that run the common validateFetchRequest
+// function which contains common validation functions
+func authorizeNodeCommon(
+	ctx context.Context,
+	storage nodeenrollment.Storage,
+	reqInfo *types.FetchNodeCredentialsInfo,
+	opt ...nodeenrollment.Option,
+) (*types.NodeInformation, error) {
+	const op = "nodeenrollment.registration.authorizeNodeCommon"
 	keyId, err := nodeenrollment.KeyIdFromPkix(reqInfo.CertificatePublicKeyPkix)
```
Comment (irenarindos) [inline registration/register_node_led.go:326]: Check for nil storage/ req?

File: types/github.com.hashicorp.nodeenrollment.types.v1.proto [hunk 2]
```diff
@@ -223,4 +223,25 @@ message GenerateServerCertificatesResponse {
...
+  // data are performed on the serialized versions of this.
+  bytes nonce = 20;
+
+  // The time at which this was created
+  google.protobuf.Timestamp creation_time = 30 [json_name="creation_time"];
+
+  // State is data that the implementor of a Store can use to round-trip data
+  // through this library; as an example, a verison number on the resource for
+  // implementing transactions.
+  google.protobuf.Struct state = 50;
```
Comment (irenarindos) [inline types/github.com.hashicorp.nodeenrollment.types.v1.proto:246]:
```suggestion
  google.protobuf.Struct state = 40;
```

**Review by irenarindos — CHANGES_REQUESTED** [review]

File: const.go [hunk 3]
```diff
@@ -32,6 +32,8 @@ const (
 	// to two weeks. Rotations should happen at roughly half this.
 	DefaultCertificateLifetime = time.Hour * 24 * 14
 
+	DefaultMaximumServerLedActivationTokenLifetime = time.Hour * 24 * 14
```
[the remark this answers is not in the document]
Comment (jefferai) [reply const.go:35]: Done

File: options_test.go [hunk 4]
```diff
@@ -156,4 +156,13 @@ func Test_GetOpts(t *testing.T) {
...
 	})
+	t.Run("with-activation-token", func(t *testing.T) {
+		assert, require := assert.New(t), require.New(t)
+		opts, err := GetOpts()
+		require.NoError(err)
+		assert.Empty(opts.WithActivationToken)
+		opts, err = GetOpts(WithActivationToken("foo"))
+		require.NoError(err)
+		assert.Equal("foo", opts.WithActivationToken)
+	})
```
[the remark this answers is not in the document]
Comment (jefferai) [reply options_test.go:167]: Done

File: registration/register_node_led.go [hunk 5]
```diff
@@ -94,18 +98,100 @@ func validateFetchRequest(
...
+	case nodeenrollment.NonceSize:
+		// Nothing; use as-is
+	default:
+		// Treat it as a server-led activation token
+		if fromAuthorize {
+			return nil, nil, fmt.Errorf("(%s) server-led activation tokens cannot be used with node-led authorize call", op)
+		}
+		consumingServerActivationToken = true
+
+		nonce := strings.TrimPrefix(string(reqInfo.Nonce), nodeenrollment.ServerLedActivationTokenPrefix)
```
[the remark this answers is not in the document]
Comment (jefferai) [reply registration/register_node_led.go:113]: Ah yes, it got removed when I switched some things around, thanks!

[the remark this answers is not in the document]
Comment (jefferai) [reply registration/register_node_led.go:113]: Actually what I did is _remove_ the prefix so I just need to remove this statement.

File: types/github.com.hashicorp.nodeenrollment.types.v1.proto (hunk 2 above)
Comment (jefferai) [reply types/github.com.hashicorp.nodeenrollment.types.v1.proto:246]: I made it 50 to match other places we use state.

File: registration/register_node_led.go (hunk 1 above)
Comment (jefferai) [reply registration/register_node_led.go:326]: See the note above -- this is either called after validation or from the validation function itself.

File: registration/register_server_led.go [hunk 6]
```diff
@@ -2,164 +2,101 @@ package registration
...
-// required so that if fields are added it is not an API change.
+// The returned values are the activation token ID (used as the ID for storage)
+// and the token itself.
 //
-// Supported options: WithWrapper (passed through to LoadRootCertificates and
-// NodeInformation.Store), WithSkipStorage (useful for tests)
-func RegisterViaServerLedFlow(
+// Supported options: WithRandomReader, WithWrapper (passed through to
+// NodeInformation.Store), WithSkipStorage (useful for tests), WithState (to
+// encode state in the activation token)
```
Comment (irenarindos) [inline registration/register_server_led.go:26]: Can this comment be updated? It looks like WithSkipStorage is being used in Boundary for controller-led so documenting why that would be used outside of testing would be useful

File: registration/register_server_led.go [hunk 7]
```diff
@@ -2,164 +2,101 @@ package registration
...
-// Note: there are currently no fields in the registration request but it is
-// required so that if fields are added it is not an API change.
+// The returned values are the activation token ID (used as the ID for storage)
+// and the token itself.
 //
-// Supported options: WithWrapper (passed through to LoadRootCertificates and
-// NodeInformation.Store), WithSkipStorage (useful for tests)
-func RegisterViaServerLedFlow(
+// Supported options: WithRandomReader, WithWrapper (passed through to
+// NodeInformation.Store), WithSkipStorage (useful for tests), WithState (to
```
Comment (jefferai) [inline registration/register_server_led.go:25]:
```suggestion
// NodeInformation.Store), WithSkipStorage, WithState (to
```

**Review by irenarindos — APPROVED** [review]

File: registration/register_server_led.go (hunk 6 above)
Comment (jefferai) [reply registration/register_server_led.go:26]: Took out the "useful for test" bit

**Pull request merged**